flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/history/HistoryServerArchiveFetcher.java

```diff
@@ -0,0 +1,244 @@
+							try {
+								FileCache.copy(jobArchive.getPath(), tmpLocalPath, false);
+
+								JsonNode archive = mapper.readTree(tmpLocalFile);
+								for (JsonNode archivePart : archive.get(ArchivedJson.ARCHIVE)) {
+									String path = archivePart.get(ArchivedJson.PATH).asText();
+									String json = archivePart.get(ArchivedJson.JSON).asText();
+
+									File target;
+									if (path.equals("/joboverview")) {
+										target = new File(webOverviewDir, jobID + JSON_FILE_ENDING);
+									} else {
+										target = new File(webDir, path + JSON_FILE_ENDING);
+									}
+
+									java.nio.file.Path parent = target.getParentFile().toPath();
+
+									try {
+										Files.createDirectories(parent);
+									} catch (FileAlreadyExistsException ignored) {
+										// odd, but OK
+									}
+
+									java.nio.file.Path targetPath = target.toPath();
+									/**
+									 * We overwrite existing files since this may be another attempt at fetching this archive.
+									 * Existing files may be incomplete/corrupt.
+									 */
+									if (Files.exists(targetPath)) {
+										Files.delete(targetPath);
+									}
+
+									Files.createFile(target.toPath());
+									try (FileWriter fw = new FileWriter(target)) {
+										fw.write(json);
+										fw.flush();
+									}
+								}
+								updateOverview = true;
+							} catch (IOException e) {
+								LOG.error("Failure while fetching/processing job archive.", e);
+								/** Make sure we attempt to fetch the archive again */
+								cachedArchives.remove(jobID);
+								/** Make sure we do not include this job in the overview */
+								Files.delete(new File(webOverviewDir, jobID + JSON_FILE_ENDING).toPath());
+								/** Clean up job files we may have created */
+								File jobDirectory = new File(webJobDir, jobID);
+								try {
+									FileUtils.deleteDirectory(jobDirectory);
+								} catch (IOException ioe) {
+									LOG.debug("Could not clean up job-directory.", ioe);
+								}
+							} finally {
+								try {
+									Files.delete(tmpLocalFile.toPath());
+								}
+								catch (NoSuchFileException e) {
+									// if the file is already gone (concurrently), we don't mind
+								}
+							}
+						}
+					}
+					if (updateOverview) {
+						updateJobOverview(webDir);
+					}
+				}
+			} catch (Exception e) {
+				LOG.error("Critical failure while fetching/processing job archives.", e);
+			}
+		}
+	}
+
+	/**
+	 * This method replicates the JSON response that would be given by the {@link CurrentJobsOverviewHandler} when
+	 * listing both running and finished jobs.
+	 *
+	 * Every job archive contains a joboverview.json file containing the same structure. Since jobs are archived on
+	 * their own however the list of finished jobs only contains a single job.
+	 *
+	 * For the display in the HistoryServer WebFrontend we have to combine these overviews.
+	 */
+	private static void updateJobOverview(File webDir) {
```

Comment:
Ok sorry! I forgot about my previous comment. What do you think about adding a handler? If it's not too much overhead, it might lead to less moving parts. We can also do this as a follow up if you don't want to do it now.